src/geometry.nim

```diff
@@ -415,31 +376,39 @@ proc newTranslation*(v: Vec3f): Translation  =
         [0.0, 0.0, 0.0,  1.0]   
     ]
 
-method apply*(traslate: Translation, vec: Vec4f): Vec4f =
-    result[0] = vec[0] + traslate.mat[0][3] * vec[3]
-    result[1] = vec[1] + traslate.mat[1][3] * vec[3]
-    result[2] = vec[2] + traslate.mat[2][3] * vec[3]
-    result[3] = vec[3]
-
-method apply*(translate: Translation, pt: Point3D): Point3D {.inline.} =
-    newPoint3D(pt.x + translate.mat[0][3], pt.y + translate.mat[1][3], pt.z + translate.mat[2][3])
-
-
-type Rotation* = object of Transformation
 
-const 
-    eX* = newVec3[float32](1, 0, 0)
-    eY* = newVec3[float32](0, 1, 0)
-    eZ* = newVec3[float32](0, 0, 1)
-
-
-proc newRotX*(angle: float32): Rotation = 
-    ## Procedure that creates a new rotation around x axis: angle is given in degrees
+proc newScaling*[T](x: T): Transformation =
+    result.kind = tkScaling
+
+    when T is SomeFloat:
+        assert x != 0, "Cannot create a new scaling Transformation with zero as factor."
+        result.mat = Mat4f.id * x
+        result.inv_mat = Mat4f.id / x
+        result.mat[3][3] = 1.0
+        result.inv_mat[3][3] = 1.0
+
+    elif T is Vec3f: 
+        result.mat = [
+            [x[0], 0.0, 0.0, 0.0], 
+            [0.0, x[1], 0.0, 0.0], 
+            [0.0, 0.0, x[2], 0.0], 
+            [0.0, 0.0,  0.0, 1.0]
+        ]
+        result.inv_mat = [
+            [1/x[0], 0.0, 0.0, 0.0], 
+            [0.0, 1/x[1], 0.0, 0.0], 
+            [0.0, 0.0, 1/x[2], 0.0], 
+            [0.0,  0.0,  0.0,  1.0]   
+        ]
+
```

Comment:
I feel like it could be safer to add an else condition if the input type differs from SomeFloat or Vec3f